File: mac-app/Sources/LinkToMac/Models/Envelope.swift
import Foundation

/// Outer wire frame. Before pairing completes, `EncryptedFrame` is skipped and
/// `Message` is sent as plain JSON; every message after that is wrapped in `EncryptedFrame`.
struct EncryptedFrame: Codable {
    let nonce: String
    let ciphertext: String
}

struct Message: Codable {
    let type: String
    let payload: JSONValue
}

// MARK: - Handshake payloads

struct HelloPayload: Codable {
    let androidPublicKey: String
    let pairingToken: String?
    let deviceToken: String?
    let deviceId: String
    let deviceName: String
    let protocolVersion: Int
}

struct HelloAckPayload: Codable {
    let status: String // "paired" | "rejected"
    let deviceToken: String?
    let macDeviceName: String
}

/// Sent unencrypted, immediately on every new connection (both first pairing and reconnect),
/// so the Mac's public key never has to be crammed into the pairing QR code — see
/// docs/PROTOCOL.md. Trust is established by the pairing token (first pairing, human-verified
/// via the QR scan) or by pinning against the previously learned key (reconnect).
struct ServerHelloPayload: Codable {
    let macPublicKey: String
    let macDeviceId: String
    let macDeviceName: String
}

// MARK: - Notification payloads

struct NotificationAction: Codable {
    let title: String
    let actionId: String
}

struct NotificationPostedPayload: Codable, Identifiable {
    let id: String
    let packageName: String
    let appName: String
    let title: String
    let text: String
    let subText: String?
    let category: String?
    let postedAt: Double
    let actions: [NotificationAction]
    let iconBase64: String?
}

struct NotificationRemovedPayload: Codable {
    let id: String
}

// MARK: - Call log + SMS payloads (Phase 2)

struct CallLogEntry: Codable, Identifiable {
    let id: String
    let number: String
    let contactName: String?
    let type: String // incoming | outgoing | missed | rejected | blocked | voicemail | unknown
    let date: Double // epoch millis
    let durationSeconds: Int
}

struct CallLogSyncPayload: Codable {
    let calls: [CallLogEntry]
}

struct SmsMessage: Codable, Identifiable, Equatable {
    let id: String
    let address: String
    let body: String
    let date: Double // epoch millis
    let isOutgoing: Bool
}

struct SmsThread: Codable, Identifiable, Equatable {
    let threadId: String
    let address: String
    let contactName: String?
    let messages: [SmsMessage]

    var id: String { threadId }
}

struct SmsSyncPayload: Codable {
    let threads: [SmsThread]
}

struct SmsSendPayload: Codable {
    let address: String
    let body: String
}

// MARK: - Contacts (Phase 7)

/// One phone number per contact — the primary/first number Android's ContactsContract reports,
/// not the full set. Enough to message/call from the Mac; a contact-detail multi-number picker
/// isn't in scope.
struct ContactEntry: Codable, Identifiable, Equatable {
    let id: String
    let name: String
    let phoneNumber: String
    let isStarred: Bool
    let email: String?
    let organization: String?
}

struct ContactsSyncPayload: Codable {
    let contacts: [ContactEntry]
}

/// Opens the phone's own dialer pre-filled with the number (`ACTION_DIAL`) — the user still has
/// to tap call on the phone. Placing a call directly (`ACTION_CALL`) needs the CALL_PHONE
/// runtime permission and would let the Mac silently dial without the phone's own confirmation;
/// that's a deliberately different (and more invasive) capability this app doesn't take on.
struct ContactsDialPayload: Codable {
    let phoneNumber: String
}

struct ContactUpdatePayload: Codable {
    let id: String
    let name: String
    let phoneNumber: String
    let isStarred: Bool
    let email: String?
    let organization: String?
}

struct ContactUpdateResultPayload: Codable {
    let id: String
    let success: Bool
    let error: String?
}

struct ContactCreatePayload: Codable {
    let name: String
    let phoneNumber: String
    let email: String?
    let organization: String?
}

struct ContactCreateResultPayload: Codable {
    let success: Bool
    let error: String?
}

struct ContactDeletePayload: Codable {
    let id: String
}

struct ContactDeleteResultPayload: Codable {
    let id: String
    let success: Bool
    let error: String?
}

// MARK: - Photo + device status payloads (Phase 3)

struct PhotoThumbnail: Codable, Identifiable {
    let id: String
    let takenAt: Double // epoch millis
    let thumbnailBase64: String
}

struct PhotoPageRequestPayload: Codable {
    let offset: Int
    let limit: Int
}

struct PhotoPagePayload: Codable {
    let photos: [PhotoThumbnail]
    let hasMore: Bool
}

struct PhotoFullRequestPayload: Codable {
    let id: String
}

struct PhotoFullPayload: Codable {
    let id: String
    let dataBase64: String
    let mimeType: String
}

struct DeviceStatusPayload: Codable {
    let batteryPercent: Int
    let isCharging: Bool
}

// MARK: - Screen mirroring payloads (Phase 4)

struct MirrorConfigPayload: Codable {
    let width: Int
    let height: Int
    let fps: Int
    let spsBase64: String
    let ppsBase64: String
}

struct MirrorStoppedPayload: Codable {
    let reason: String // requested | permission_denied | error
}

struct MirrorTapPayload: Codable {
    let x: Double // normalized 0.0-1.0
    let y: Double
}

struct MirrorSwipePayload: Codable {
    let startX: Double
    let startY: Double
    let endX: Double
    let endY: Double
    let durationMs: Int
}

struct MirrorKeyPayload: Codable {
    let action: String // back | home | recents
}

struct MirrorTextInputPayload: Codable {
    let text: String
}

// MARK: - Shared clipboard (Phase 5)

struct ClipboardUpdatePayload: Codable {
    let text: String
    let sourceDeviceId: String
    let timestamp: Double // epoch millis
}

// MARK: - File browsing (Phase 6)

struct FileEntry: Codable, Identifiable, Equatable {
    let name: String
    let isDirectory: Bool
    let sizeBytes: Double
    let modifiedAt: Double // epoch millis

    var id: String { name }
}

/// `path` is always relative to the phone's shared storage root ("" is that root itself) —
/// never an absolute filesystem path, so the wire protocol never leaks device-specific mount
/// points. Segments are joined with "/"; see docs/PROTOCOL.md.
struct FilesListRequestPayload: Codable {
    let path: String
}

struct FilesListResultPayload: Codable {
    let path: String
    let entries: [FileEntry]
    let error: String?
}

struct FilesDownloadRequestPayload: Codable {
    let path: String
}

struct FilesDownloadResultPayload: Codable {
    let path: String
    let name: String
    let dataBase64: String?
    let mimeType: String?
    let error: String?
}

struct FilesUploadPayload: Codable {
    let path: String // destination directory
    let name: String
    let dataBase64: String
    let mimeType: String
}

struct FilesUploadResultPayload: Codable {
    let path: String
    let name: String
    let success: Bool
    let error: String?
}

struct FilesCreateFolderPayload: Codable {
    let path: String // parent directory
    let name: String
}

struct FilesCreateFolderResultPayload: Codable {
    let path: String
    let name: String
    let success: Bool
    let error: String?
}

struct FilesRenamePayload: Codable {
    let path: String
    let newName: String
}

struct FilesRenameResultPayload: Codable {
    let path: String
    let newName: String
    let success: Bool
    let error: String?
}

struct FilesDeletePayload: Codable {
    let path: String
}

struct FilesDeleteResultPayload: Codable {
    let path: String
    let success: Bool
    let error: String?
}

/// Shared by files.copy/files.move — both take a source item and a destination directory to
/// place it in (named after the source's own basename); they differ only in whether the source
/// survives the operation.
struct FilesTransferPayload: Codable {
    let sourcePath: String
    let destinationPath: String
}

struct FilesTransferResultPayload: Codable {
    let sourcePath: String
    let destinationPath: String
    let success: Bool
    let error: String?
}

/// Minimal untyped JSON box so `Message.payload` can hold any of the payload structs above
/// without a giant enum of coding keys. Encoded/decoded via JSONSerialization under the hood.
enum JSONValue: Codable {
    case object([String: JSONValue])
    case array([JSONValue])
    case string(String)
    case number(Double)
    case bool(Bool)
    case null

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let value = try? container.decode([String: JSONValue].self) {
            self = .object(value)
        } else if let value = try? container.decode([JSONValue].self) {
            self = .array(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode(Double.self) {
            self = .number(value)
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else {
            self = .null
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .object(let value): try container.encode(value)
        case .array(let value): try container.encode(value)
        case .string(let value): try container.encode(value)
        case .number(let value): try container.encode(value)
        case .bool(let value): try container.encode(value)
        case .null: try container.encodeNil()
        }
    }

    /// Round-trips through a `Codable` struct.
    func decoded<T: Decodable>(as type: T.Type) throws -> T {
        let data = try JSONEncoder().encode(self)
        return try JSONDecoder().decode(T.self, from: data)
    }

    static func from<T: Encodable>(_ value: T) throws -> JSONValue {
        let data = try JSONEncoder().encode(value)
        return try JSONDecoder().decode(JSONValue.self, from: data)
    }
}

File: desktop-app/src-tauri/src/protocol/envelope.rs
//! Wire-protocol structs, ported 1:1 from `mac-app/Sources/LinkToMac/Models/Envelope.swift`.
//! See `docs/PROTOCOL.md` for the authoritative spec. `serde_json::Value` stands in for the
//! Swift app's hand-rolled `JSONValue` enum — it does the same job natively in Rust.

use serde::{Deserialize, Serialize};

/// Outer wire frame once a session key exists. Before pairing completes, `Message` is sent
/// as plain JSON with no `EncryptedFrame` wrapper.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EncryptedFrame {
    pub nonce: String,
    pub ciphertext: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    #[serde(rename = "type")]
    pub message_type: String,
    pub payload: serde_json::Value,
}

// MARK: - Handshake payloads

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HelloPayload {
    pub android_public_key: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pairing_token: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub device_token: Option<String>,
    pub device_id: String,
    pub device_name: String,
    pub protocol_version: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HelloAckPayload {
    /// "paired" | "rejected"
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub device_token: Option<String>,
    pub mac_device_name: String,
}

/// Sent unencrypted, immediately on every new connection (both first pairing and reconnect) —
/// see docs/PROTOCOL.md. Trust is established by the pairing token (first pairing) or by
/// pinning against the previously learned key (reconnect).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServerHelloPayload {
    pub mac_public_key: String,
    pub mac_device_id: String,
    pub mac_device_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EmptyPayload {}

/// The pairing QR's contents — deliberately doesn't include the Mac's public key (see
/// `serverHello`'s doc comment for why); ported from `PairingQRPayload` in
/// `ConnectionServer.swift`. Not part of the encrypted-envelope wire protocol itself, but sent
/// to the frontend so it can render the QR code and to Android by scanning it.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PairingQrPayload {
    pub host: String,
    pub port: u16,
    #[serde(rename = "pairingToken")]
    pub pairing_token: String,
    #[serde(rename = "macDeviceId")]
    pub mac_device_id: String,
}

// MARK: - Notification payloads

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NotificationAction {
    pub title: String,
    pub action_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotificationPostedPayload {
    pub id: String,
    pub package_name: String,
    pub app_name: String,
    pub title: String,
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sub_text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
    pub posted_at: f64,
    pub actions: Vec<NotificationAction>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon_base64: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NotificationRemovedPayload {
    pub id: String,
}

// MARK: - Call log + SMS payloads (Phase 2)

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallLogEntry {
    pub id: String,
    pub number: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub contact_name: Option<String>,
    /// incoming | outgoing | missed | rejected | blocked | voicemail | unknown
    #[serde(rename = "type")]
    pub call_type: String,
    pub date: f64,
    pub duration_seconds: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CallLogSyncPayload {
    pub calls: Vec<CallLogEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SmsMessage {
    pub id: String,
    pub address: String,
    pub body: String,
    pub date: f64,
    pub is_outgoing: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SmsThread {
    pub thread_id: String,
    pub address: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub contact_name: Option<String>,
    pub messages: Vec<SmsMessage>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SmsSyncPayload {
    pub threads: Vec<SmsThread>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SmsSendPayload {
    pub address: String,
    pub body: String,
}

// MARK: - Contacts (Phase 7)

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContactEntry {
    pub id: String,
    pub name: String,
    pub phone_number: String,
    pub is_starred: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub organization: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactsSyncPayload {
    pub contacts: Vec<ContactEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContactsDialPayload {
    pub phone_number: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContactUpdatePayload {
    pub id: String,
    pub name: String,
    pub phone_number: String,
    pub is_starred: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub organization: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactUpdateResultPayload {
    pub id: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContactCreatePayload {
    pub name: String,
    pub phone_number: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub organization: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactCreateResultPayload {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactDeletePayload {
    pub id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactDeleteResultPayload {
    pub id: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

// MARK: - Photo + device status payloads (Phase 3)

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PhotoThumbnail {
    pub id: String,
    pub taken_at: f64,
    pub thumbnail_base64: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PhotoPageRequestPayload {
    pub offset: i64,
    pub limit: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PhotoPagePayload {
    pub photos: Vec<PhotoThumbnail>,
    pub has_more: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PhotoFullRequestPayload {
    pub id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PhotoFullPayload {
    pub id: String,
    pub data_base64: String,
    pub mime_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeviceStatusPayload {
    pub battery_percent: i64,
    pub is_charging: bool,
}

// MARK: - Screen mirroring payloads (Phase 4)

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MirrorConfigPayload {
    pub width: i64,
    pub height: i64,
    pub fps: i64,
    pub sps_base64: String,
    pub pps_base64: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MirrorStoppedPayload {
    /// requested | permission_denied | error
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MirrorTapPayload {
    /// normalized 0.0-1.0
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MirrorSwipePayload {
    pub start_x: f64,
    pub start_y: f64,
    pub end_x: f64,
    pub end_y: f64,
    pub duration_ms: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MirrorKeyPayload {
    /// back | home | recents
    pub action: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MirrorTextInputPayload {
    pub text: String,
}

// MARK: - Shared clipboard (Phase 5)

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClipboardUpdatePayload {
    pub text: String,
    pub source_device_id: String,
    pub timestamp: f64,
}

// MARK: - File browsing (Phase 6)

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileEntry {
    pub name: String,
    pub is_directory: bool,
    pub size_bytes: f64,
    pub modified_at: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesListRequestPayload {
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesListResultPayload {
    pub path: String,
    pub entries: Vec<FileEntry>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesDownloadRequestPayload {
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FilesDownloadResultPayload {
    pub path: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data_base64: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FilesUploadPayload {
    /// destination directory
    pub path: String,
    pub name: String,
    pub data_base64: String,
    pub mime_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesUploadResultPayload {
    pub path: String,
    pub name: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesCreateFolderPayload {
    /// parent directory
    pub path: String,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesCreateFolderResultPayload {
    pub path: String,
    pub name: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FilesRenamePayload {
    pub path: String,
    pub new_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FilesRenameResultPayload {
    pub path: String,
    pub new_name: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesDeletePayload {
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesDeleteResultPayload {
    pub path: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Shared by files.copy/files.move — both take a source item and a destination directory.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FilesTransferPayload {
    pub source_path: String,
    pub destination_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FilesTransferResultPayload {
    pub source_path: String,
    pub destination_path: String,
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}
